File: src/main/java/akka/actor/AggregateActor.java
package akka.actor;

import akka.message.ReduceDataResult;
import akka.message.Result;

import java.util.HashMap;
import java.util.Map;

/**
 * @user sun
 * @date 2016/7/4
 */
public class AggregateActor extends UntypedActor {

    private Map<String, Integer> finalReduceMap = new HashMap<>();

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof ReduceDataResult) {
            ReduceDataResult dataResult = (ReduceDataResult) message;
            aggregateInMemoryReduce(dataResult.getReduceDataMap());
        } else if (message instanceof Result) {
            System.out.println(finalReduceMap.toString());
        } else
            unhandled(message);
    }

    private void aggregateInMemoryReduce(Map<String, Integer> reduceMap) {
        Integer count;
        for (String key : reduceMap.keySet()) {
            if (finalReduceMap.containsKey(key)) {
                count = reduceMap.get(key) + finalReduceMap.get(key);
                finalReduceMap.put(key, count);
            } else {
                finalReduceMap.put(key, reduceMap.get(key));
            }
        }
    }
}

File: src/main/scala/scala_akka/actor/AggregateActor.scala
package scala_akka.actor

import java.util

import akka.actor.Actor

import scala_akka.message.{ReduceDataResult, Result}

/**
  * @user sun 
  * @date 2016/7/10
  */
class AggregateActor extends Actor {
  var finalReduceMap = new util.HashMap[String, Integer]

  override def receive: Receive = {
    case message: ReduceDataResult =>
      aggregateInMemoryReduce(message.reduceMap)
    case message: Result =>
      System.out.println(finalReduceMap.toString)
    case _ =>
  }

  def aggregateInMemoryReduce(reduceMap: util.HashMap[String, Integer]) = {
    var count: Integer = 0
    import scala.collection.JavaConversions._
    val sets: util.Set[String] = reduceMap.keySet()
    for (key <- sets) {
      if (finalReduceMap.containsKey(key)) {
        count = reduceMap.get(key) + finalReduceMap.get(key)
        finalReduceMap.put(key, count)
      } else {
        finalReduceMap.put(key, reduceMap.get(key))
      }
    }
  }
}
